### PR TITLE
Add hinge loss op

### LINKED ISSUE #5757: Hinge Loss

### PR DESCRIPTION
**Pull request by sidgoyal78**
Resolves #5757 by adding an implementation of a hinge loss operator for a binary classification scenario.

The hinge loss for a prediction (x) and a target label (y), where x can be in (-inf, inf) and y can be +1 or -1, is:

```
hloss(x, y) = max(1 - y.x, 0) 
```
Future work:
- To extend this to multi-class setting
- More efficient implementation

### CODE REVIEW

File: paddle/operators/hinge_loss_op.cc [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+ public:
+  HingeLossOpMaker(framework::OpProto* proto,
+                   framework::OpAttrChecker* op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("Logits",
+             "The input value (Logits) of Hinge loss op."
+             "Logits is a 2-D tensor with shape [batch_size, 1].");
+    AddInput("Labels",
+             "The target value (Labels) of Hinge loss op."
+             "Labels is a 2-D tensor with shape [batch_size, 1].");
```
[the remark this answers is not in the document]
Comment (qingqing01) [reply paddle/operators/hinge_loss_op.cc]: In the implementation of Caffe: https://github.com/BVLC/caffe/blob/master/src/caffe/layers/hinge_loss_layer.cpp#L24

The dimension of `Logits` is not 1.

File: paddle/operators/hinge_loss_op.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    auto* loss = context.Output<framework::Tensor>("Loss");
+    auto place = context.GetEigenDevice<Place>();
+
+    auto x = framework::EigenVector<T>::Flatten(*pred);
+    auto y = framework::EigenVector<T>::Flatten(*label);
+    loss->mutable_data<T>(context.GetPlace());
+    auto l = framework::EigenVector<T>::Flatten(*loss);
+    l.device(place) =
+        (static_cast<T>(1) - x * (static_cast<T>(2) * y - static_cast<T>(1)))
+            .cwiseMax(static_cast<T>(0));
```
Comment (qingqing01) [inline paddle/operators/hinge_loss_op.h]: Why the implementation here is different from Caffe: https://github.com/BVLC/caffe/blob/master/src/caffe/layers/hinge_loss_layer.cpp#L10  ?

File: paddle/operators/hinge_loss_op.cc [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    AddComment(R"DOC(
+HingeLoss Operator.
+
+Let x be a logit (prediction) and y be the actual label. The logit can
+take any values from (-inf, inf), but the labels should be either -1 or 1.
+Then, the hinge loss is computed as follows:
+
+L_(x, y) = max(1 - y.x, 0) 
+
+Note that the labels passed as input will have values as either 0 or 1.
```
Comment (qingqing01) [inline paddle/operators/hinge_loss_op.cc]: Is there any reference paper?

File: paddle/operators/hinge_loss_op.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (sidgoyal78) [reply paddle/operators/hinge_loss_op.cc]: Yes, the width of `Logits` should also be 1.
I think in caffe, the implementation is for a multi-class classification setting. (Reference: http://caffe.berkeleyvision.org/doxygen/classcaffe_1_1HingeLossLayer.html)

The implementation here is for a standard binary classification task. For instance, please see: tensorflow's implementation: https://github.com/tensorflow/tensorflow/blob/r1.4/tensorflow/python/ops/losses/losses_impl.py

File: paddle/operators/hinge_loss_op.cc (hunk 3 above)
Comment (sidgoyal78) [reply paddle/operators/hinge_loss_op.cc]: I didn't find the original paper, but here is one: http://web.mit.edu/lrosasco/www/publications/loss.pdf (page 6).
If wiki reference is okay, then maybe this: https://en.wikipedia.org/wiki/Loss_functions_for_classification#Hinge_loss

File: paddle/operators/hinge_loss_op.cc [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+              "The output tensor with shape [batch_size, 1] "
+              "which represents the hinge loss.");
+    AddComment(R"DOC(
+HingeLoss Operator.
+
+Let x be a logit (prediction) and y be the actual label. The logit can
+take any values from (-inf, inf), but the labels should be either -1 or 1.
+Then, the hinge loss is computed as follows:
+
+L_(x, y) = max(1 - y.x, 0)
```
[the remark this answers is not in the document]
Comment (sidgoyal78) [reply paddle/operators/hinge_loss_op.cc]: For second point, I think that it is under the assumption that the `Label` tensor will have values which are either 0 or 1. However, for the computation of the loss, we need labels to be either -1 or 1. Hence, i do the transformation.

File: paddle/operators/hinge_loss_op.h (hunk 2 above)
Comment (sidgoyal78) [reply paddle/operators/hinge_loss_op.h]: I think that implementation is for multi-class (please see link above).

File: paddle/operators/hinge_loss_op.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (qingqing01) [reply paddle/operators/hinge_loss_op.cc]: Oh, I see. Thanks!

File: paddle/operators/hinge_loss_op.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (qingqing01) [reply paddle/operators/hinge_loss_op.cc]: Oh, I see. Thanks!
